doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+
+// User provides data and offsets as separate inputs
+auto tv_data = TensorViewBuilder()
+    .ndims(2)
+    .shape({-1, 512})  // [total_tokens, hidden]
+    .dtype(DataType::Float)
+    .build();
+fusion.addInput(tv_data);
+
+auto tv_offsets = TensorViewBuilder()
+    .ndims(1)
+    .shape({9})  // [num_experts + 1]
+    .dtype(DataType::Int)
+    .build();
+fusion.addInput(tv_offsets);
+
+// User explicitly creates nested tensor
+auto tv_nested = asNested(tv_data, tv_offsets, /*ragged_dim=*/0);
+// tv_nested has shape [batch=8, ragged_tokens, hidden=512]
+
+// Operations on the nested tensor
+auto tv_result = some_operation(tv_nested);
+
+fusion.addOutput(tv_result);
+```
+
+The output tensor, `tv_result`, is a nested tensor. The extents of the nested domains are given as a fusion input, but in general, they are not known until the fusion is executed. Thus, if the nested tensor struct were defined like:
+
+```cpp
+template <typename DT, int rank>
+struct NestedTensor {
+	DT* ptr;
+	int64_t extents[rank];
+	int64_t nested_domain_extents[ragged_dimension_rank];
+};
+```
+
+The value of `nested_domain_extents` is not available until the completion of the kernel, which would block the launch of the subsequent kernel.
+
+Instead, we would like the fusion to be defined as follows:
+
+```cpp
+fusion.addInput(tv_data);      // Original data input (unchanged)
+fusion.addInput(tv_offsets);   // Original offset input (unchanged)
+
+auto tv_nested = asNested(tv_data, tv_offsets, /*ragged_dim=*/0);
+auto tv_result = some_operation(tv_nested);
+
+auto tv_result_offsets = /* extract/compute offset part of tv_result */;
+
+fusion.addOutput(tv_result);      // Data tensor output
+fusion.addOutput(tv_result_offsets);   // Offset tensor output (injected)
+```
+
+Here, for `tv_result` we would use the same `Tensor` struct as the normal tensor. The offset tensor would be a 1D tensor with the `ptr` val referring to the vector holding the offsets on the device memory. In this case, there's nothing to block the launch of the subsequent kernel as the offset vector would remain on the device memory.
+
+Since it is an implementation detail, the offset tensor should be hidden behind the nested tensor in the user-facing Fusion definition. When a user uses `asNested` to create a nested tensor, it should still create a single nested tensor Val, as illustrated in the first case above. The translation to the second pattern should be done automatically, e.g., by a new preseg pass.
+
+#### Runtime Binding (Initial Version)
+
+**Challenge**: How should `ExpressionEvaluator` bind a ragged TensorView to actual tensor objects at runtime?
+
+**Initial Approach**: In this initial version, we take a simplified approach that avoids the complexity of custom tensor types:
+
+1. **Input Nested Tensors**: Must be passed as **two separate parameters**:
+   - A normal `at::Tensor` containing the flattened data
+   - Another `at::Tensor` parameter representing the offsets (or extents)
```

Comment:
tokens_per_expert and tokens_per_expert_per_device are actual TensorViews in the graph so they'll be bound to extent tensors anyway. A nested TensorView (i.e. a TensorView whose logical domain contains a RaggedIterDomain) doesn't need to be bound to two tensors. I think this will simplify the implementation a lot.

Reply:
> bind a ragged TensorView to actual tensor objects at runtime?

By this, don't you mean **one** TensorView is bound to **two** tensor objects? That's what I was referring to and disagree.

Reply:
Indirectly, yes. The RaggedIterDomain will be linked to the tensor object for the offsets. I don't think if we could do without having a link to the offset tensor from the ragged iter domain. We would need to maintain some connection to find its offsets.

Reply:
> Indirectly, yes. 

Agreed. 

I asked this question before knowing your plan to first hide ragged TensorVies from ExpressionEvaluator. Since I understood that, eventually, we should still maintain one-to-one between TVs and tensors. For example,
```
TV: [e, j1, h] => tensor: [s, h]
TV: j1's extent => tensor: [e]
```
Correct?

Reply:
Right, I just meant there's no single Aten type that could be used to represent an nvFuser nested tensor. As we discussed before, PyTorch's nested tensor is not sufficient for us as it would mean offsets are always copied between host and device.